### DISCUSSION TITLE
How do I handle difference responses to the same endpoint?

### DISCUSSION THREAD

**Discussion by CodyBontecou** [mswjs/msw#885] [Q&A]
I'd like to test views when given no data, errors, and expected data.

I can create the handler to manage a single response, but what's the best way to manage multiple responses for the same endpoint depending on the state I'm trying to test?

**Comment by CodyBontecou** [comment]
I found the [permanent override](https://mswjs.io/docs/api/setup-worker/use#permanent-override). Is this the proper solution? Write my handlers within the test file itself?

**Reply by msutkowski** [reply]
@CodyBontecou I like to set handlers for default behaviors, and then for special cases leverage a combination of `res.once` and individual handlers like you're saying on a test-by-test basis.

**Comment by kettanaito** [comment]
Hey, @CodyBontecou.

Please use the [runtime handlers](https://mswjs.io/docs/api/setup-server/use) in the combination with [resetting handlers between tests](https://mswjs.io/docs/api/setup-server/reset-handlers). That way you don't have to use `res.once()`.

```js
const server = setupServer(
  rest.get('/resource', (req, res, ctx) => res(ctx.text('hello')))
)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

test('happy path', ()  => {
  // make request.
})

test('error stat', () => {
  server.use(
    rest.get('/resource', (req, res, ctx) => res(ctx.status(500)))
  )
  // make request
})

test('another state', () => {
  server.use(
    // Model any response overrides you need.
    rest.get('/resource', (req, res, ctx) => res(ctx.status(403)))
  )
  // make request
})
```

This setup is also illustrated in the [React Testing Library example](https://testing-library.com/docs/react-testing-library/example-intro).

**Reply by CodyBontecou** [reply]
Do you use setupWorker when working with Cypress? 

Does setupWorker provide the server.use method to override the default requests?

**Reply by kettanaito** [reply]
Yes, I'd recommend using `setupWorker` with Cypress.

Yes, there's a [`worker.use()`](https://mswjs.io/docs/api/setup-worker/use) method.

**Reply by CodyBontecou** [reply]
Could you please look over [this file](https://github.com/CodyBontecou/msw-cypress-nextjs-typescript/blob/main/cypress/integration/home.spec.js ) give me pointers on why the mocks aren't happening?

**Reply by kettanaito** [reply]
You’re initiating a worker in a test file which isn’t the right place to do so. I highly recommend following the common browser setup from the docs and accessing worker from cy.window() as shown here: https://mswjs.io/docs/api/setup-worker/use#runtime-request-handler

**Reply by CodyBontecou** [reply]
Hmm, interesting. What's the best way to attach msw to the window object within a Next.js project alongside Cypress?

**Reply by kettanaito** [reply]
Set it on the `window` object where you create your worker.

```js
const worker = setupWorker(...handlers)

window.worker = worker
```

Cypress still runs your application, so whatever you set on `window` you can access in Cypress via `cy.window()`.

**Reply by CodyBontecou** [reply]
Okay, I have access to window.msw within my local web browser, but it appears to be undefined when running Cypress.

**Reply by kettanaito** [reply]
Submit a reproduction repository then so we could take a look.

**Reply by CodyBontecou** [reply]
Sure, here you go https://github.com/CodyBontecou/msw-cypress-nextjs-typescript

**Reply by CodyBontecou** [reply]
Hey @kettanaito, just a friendly bump. Is this repo explicit enough, or would you like me to break it down further to the specific problem?

**Reply by kettanaito** [reply]
I haven't got time to look into this. I'll update this thread once I do.
